ua-snap/hydroviz
# Add encodings and build ingest JSON

This PR is intended to make the assembled netCDF data more `rasdaman`-friendly. It builds the netCDF files with integer encodings for the dimensions instead of strings, and also adds a bunch of useful attributes to the netCDF files. Some of these attributes are the encodings themselves, which allows us to build an ingest JSON directly from the netCDF file (ie, not using any auxillary files to crosswalk the integer encodings to the ingest recipe JSON file). The ingest JSONs are created programatically using a new `build_injest_json.py` script, which runs immediately after netCDF file generation. 

To test:

- Clone the repo somewhere on Chinook, and make sure you have the `snap-geo` environment built.
- Follow the instructions in the `README.md` to run the code in the _"To coerce the data into a netCDF format..."_ bullet, saving the outputs to a different directory than in the example.
- Check dimensions in netCDF outputs. View the netCDF files using `ncdump -h` or in a notebook, and make sure all dimensions are encoded with integers.
- Check out the netCDF attributes. Is this too much / too messy? I'd like to include as much metadata as possible in the files themselves, since the information is rather scattered in the source data. Including the encodings crosswalk seems like good practice here, but let me know if there is a reason to avoid it, or if you think `rasdaman` will balk at this for some reason.
- Check out the ingest JSON files (`hru_ingest.json` and `seg_ingest.json`). **These will need work!** These were modeled after the ingest [here](https://github.com/ua-snap/rasdaman-ingest/blob/7536c17a0df2ecec18c965f030d7e943bccd2d01/iem/gipl2_4km/ingest.json), but as you can see I have removed the WMS hooks and the X and Y axes. I kept the CRS information in the config and recipe sections (assuming a CRS must be defined?) but there is no spatial information in this dataset and I am unsure what needs to be included or excluded. As @kyleredilla said "this is uncharted waters", plus I have 0% experience here, so please let me know if you have any ideas on how to format this ingest for a non-spatial netCDF!
